parsers/BD.py

```diff
@@ -1,299 +1,295 @@
+        # [3] is total demand in MW
+        row_dict["total_demand"] = table_entry_to_float(row_items[3])
+        # [4] is loadshed, not interesting for us
+        row_dict["loadshed"] = table_entry_to_float(row_items[4])
+        # [5] is generation from gas
+        row_dict["gas"] = table_entry_to_float(row_items[5])
+        # [6] is generation from liquid fuel (oil)
+        row_dict["oil"] = table_entry_to_float(row_items[6])
+        # [7] is generation from coal
+        row_dict["coal"] = table_entry_to_float(row_items[7])
+        # [8] is generation from hydro
+        row_dict["hydro"] = table_entry_to_float(row_items[8])
+        # [9] is generation from solar
+        row_dict["solar"] = table_entry_to_float(row_items[9])
+        # [10], [11] is import from Bheramara PP (IN-EA) and Tripura (IN-NE)
+        row_dict["bd_import_bheramara"] = table_entry_to_float(row_items[10])
+        row_dict["bd_import_tripura"] = table_entry_to_float(row_items[11])
+        row_dict["remarks"] = row_items[12]
+
+        row_data.append(row_dict)
+
+    return row_data
+
+
+def verify_table(table_header: Tag):
+    """
+    Validate the table headers, that it looks like expected.
+    Don't parse if the table has changed.
+    """
+    header_items = table_header.find_all("th")
+    header_items = [item.text.strip() for item in header_items]
 
-    return processed_data
+    if header_items != TABLE_HEADERS:
+        raise ParserException(
+            parser="BD.py",
+            message=(
+                f"Table headers mismatch with expected ones."
+                f"Expected: {TABLE_HEADERS}"
+                f"Parsed: {header_items}"
+            ),
+        )
 
 
-def exchange_processer(df, target_datetime, old_format=False) -> list:
+def query(
+    session: Optional[Session], target_datetime: Optional[datetime], logger: Logger
+) -> List[Dict[str, Any]]:
     """
-    Takes dataframe and extracts all exchange data and timestamps.
-    Returns a list of 2 element tuples in form (dict, arrow object).
+    Query the table and read it into list.
     """
+    # build URL to call
+    if target_datetime is None:
+        target_url = latest_url
+    else:
+        # Convert timezone of target_datetime, and build URL from day
+        target_datetime_bd = arrow.get(target_datetime).to(tz)
+        target_dt_str = target_datetime_bd.format("DD-MM-YYYY")
+        target_url = url_by_date + target_dt_str
 
-    # Positive means import from India hence sign reversal needed for EM.
-    processed_data = []
-    for index, row in df.iterrows():
-        exchange = row.to_dict()
-        flow = exchange.pop("HVDC", 0.0) + exchange.pop("Tripura", 0.0)
-        dt = timestamp_converter(index, target_datetime)
-        processed_data.append((-1 * flow, dt))
+    target_response: Response = session.get(target_url)
 
-    return processed_data
+    if not target_response.ok:
+        return_status = target_response.status_code
```

Comment:
We can probably avoid this varable declaration and just include the target_response.status_code directly in the message instead.

Reply:
Done.